File: python_solutions/111.minimum-depth-of-binary-tree.py
#
# @lc app=leetcode id=111 lang=python
#
# [111] Minimum Depth of Binary Tree
#
# https://leetcode.com/problems/minimum-depth-of-binary-tree/description/
#
# algorithms
# Easy (34.51%)
# Total Accepted:    260.3K
# Total Submissions: 754K
# Testcase Example:  '[3,9,20,null,null,15,7]'
#
# Given a binary tree, find its minimum depth.
#
# The minimum depth is the number of nodes along the shortest path from the
# root node down to the nearest leaf node.
#
# Note: A leaf is a node with no children.
#
# Example:
#
# Given binary tree [3,9,20,null,null,15,7],
#
#
# ⁠   3
# ⁠  / \
# ⁠ 9  20
# ⁠   /  \
# ⁠  15   7
#
# return its minimum depth = 2.
#
#
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        left = self.minDepth(root.left)
        right = self.minDepth(root.right)
        return 1 + (left + right if left == 0 or right ==
                    0 else min(left, right))


# print(Solution().minDepth(TreeNode(0)))

File: _site/solutions/111_minimum_depth_of_binary_tree.cpp
#include<stdio.h>
#include<iostream>
#include<vector>
#include<algorithm>
#include<climits>
/**
   Given a binary tree, find its minimum depth.
   The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.
   Note:&nbsp;A leaf is a node with no children.
   Example:
   Given binary tree [3,9,20,null,null,15,7],
   3
   / \
   9  20
   /  \
   15   7
   return its minimum&nbsp;depth = 2.

**/
using namespace std; 

static int speed_up = []() { std::ios::sync_with_stdio(false); cin.tie(NULL); return 0; }(); 


// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  int minDepth(TreeNode* root) {
    if (!root) return 0;
    if (!root->left) return 1 + minDepth(root->right);
    if (!root->right) return 1 + minDepth(root->left);
    return 1 + min(minDepth(root->left), minDepth(root->right));
  }
};

int main() { 
  Solution s;
}
